src/java.base/share/classes/java/util/zip/ZipInputStream.java

```diff
@@ -515,9 +515,15 @@ private ZipEntry readLOC() throws IOException {
         }
         readFully(b, 0, len);
         // Force to use UTF-8 if the USE_UTF8 bit is ON
-        ZipEntry e = createZipEntry(((flag & USE_UTF8) != 0)
-                                    ? ZipCoder.toStringUTF8(b, len)
-                                    : zc.toString(b, len));
+        ZipEntry e;
+        try {
+            e = createZipEntry(((flag & USE_UTF8) != 0)
+                    ? ZipCoder.toStringUTF8(b, len)
+                    : zc.toString(b, len));
+        } catch(Exception ex) {
+            throw  (ZipException) new ZipException(
+                    "invalid LOC header (bad entry name)").initCause(ex);
+        }
```

Comment:
There is a lot going on here:

* The creation of a ZipEntry
* The interpretation of the general purpose bit flag
* The split to call either a static or non-static toString depending on the flag interpretation
* The exception handling, including the somewhat unusual initCause and cast

The comment at 517 seems even more detached from the logic it tries to describe after this change.

Perhaps we could benefit from introducing a `zipCoderFromFlag(int flag)` method, similar to that in ZipFile?

This would allow the section to be rewritten to something that hide the flag parsing details, leaves us with a single (polymorphic) toString invocation and extracts `createZipEntry` from the try/catch scope. 

Something like this: (I also changed the exception handling here, that's a bit orthogonal to the main points above)

```suggestion
        String name;
        try {
            name = zipCoderForFlag(flag).toString(b, len);
        } catch (IllegalArgumentException ex) {
            ZipException e = new ZipException("invalid LOC header (bad entry name)");
            e.initCause(ex);
            throw e;
        }
        ZipEntry e = createZipEntry(name);
```

Reply:
I believe you meant `ZipCoderFromPos(int flag)`.  I don't think it is needed in ZipInputStream but I did move the call to createZipEntry out of the try block per your suggestion